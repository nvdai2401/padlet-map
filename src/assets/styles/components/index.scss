@import './top-bar';
@import './folk-me';
@import './spinner';
@import './post-content';
@import './post-popup';
@import './mini-post';
@import './post-preview';
@import './post-expanded';
@import './map-control';
@import './map-view';
