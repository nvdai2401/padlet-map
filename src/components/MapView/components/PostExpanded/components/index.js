export { default as PostSlider } from './PostSlider';
