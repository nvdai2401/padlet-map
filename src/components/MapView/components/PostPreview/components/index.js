export { default as MiniPost } from './MiniPost';
