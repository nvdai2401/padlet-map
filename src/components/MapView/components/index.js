export { default as Post } from './Post';
export { default as PostPreview } from './PostPreview';
export { default as PostExpanded } from './PostExpanded';
