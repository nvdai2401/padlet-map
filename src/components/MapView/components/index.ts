export { default as PostContent } from './PostContent';
export { default as PostPopup } from './PostPopup';
export { default as PostPreview } from './PostPreview';
export { default as PostExpanded } from './PostExpanded';
export { default as MapControl } from './MapControl';
