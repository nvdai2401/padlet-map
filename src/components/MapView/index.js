export { default } from './MapView.vue';
