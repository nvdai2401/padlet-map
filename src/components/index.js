export { default as MapView } from './MapView';
export { default as TopBar } from './TopBar';
