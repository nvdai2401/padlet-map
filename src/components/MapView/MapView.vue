<template>
  <div class="map-view">
    <gmap-map
      ref="gmap"
      class="gmap-container"
      :zoom="initialZoom"
      :center="center"
      :options="mapOptions"
      @click="handleOnClickMap"
    >
      <gmap-marker
        ref="markers"
        v-for="marker in Object.values(markers)"
        :key="marker.id"
        :icon="{
          url: require(`src/assets/images/markers/${marker.color}_marker.webp`),
        }"
        :clickable="true"
        :position="{
          lat: marker.location_point.latitude,
          lng: marker.location_point.longitude,
        }"
        @click="handleOnClick(marker.id)"
        @mouseover="handleOnMouseOver(marker.id)"
        @mouseout="handleOnMouseOut(marker.id)"
      >
        <gmap-info-window
          :opened="marker.is_post_visible"
          @closeclick="is_post_visible = false"
        >
          <post
            v-if="marker.is_post_visible"
            :post-info="marker"
            :onExpandPost="handleOnExpandPost"
          />
        </gmap-info-window>
      </gmap-marker>
    </gmap-map>

    <div :class="['map-control', postPreviewVisible ? 'preview-shown' : '']">
      <button
        class="button map-center-control"
        @click="postPreviewVisible = !postPreviewVisible"
      >
        <font-awesome-icon :icon="['fas', 'bars']" />
      </button>

      <div class="group-button">
        <button class="button" @click="fitBounds">
          <font-awesome-icon :icon="['fas', 'expand']" />
        </button>
        <button class="button" @click="zoomIn">
          <font-awesome-icon :icon="['fas', 'plus']" />
        </button>
        <button class="button" @click="zoomOut">
          <font-awesome-icon :icon="['fas', 'minus']" />
        </button>
      </div>
    </div>
    <transition-group tag="div" name="slide">
      <post-preview
        v-show="postPreviewVisible && Object.keys(markers).length > 0"
        key="post-preview"
        :posts="markers"
        :onClick="handleOnClick"
        :onMouseOver="handleOnMouseOver"
        :onMouseOut="handleOnMouseOut"
      />
    </transition-group>

    <post-expanded
      v-if="postExpandedVisible"
      :open="postExpandedVisible"
      :postId="activePost"
      :posts="Object.keys(markers)"
      :onClose="handleOnClosePostExpanded"
    />
  </div>
</template>

<script>
import { Post, PostPreview, PostExpanded } from "./components";
import {
  defaultLatLngBounds,
  viewportLatLngBounds,
  options,
  DEFAULT_CENTER,
  MIN_ZOOM,
  MAX_ZOOM,
} from "./mapProperty";
import { getMarkers } from "src/api";

export default {
  name: "map-view",
  components: {
    post: Post,
    "post-preview": PostPreview,
    "post-expanded": PostExpanded,
  },
  data() {
    return {
      initialZoom: MIN_ZOOM,
      center: DEFAULT_CENTER,
      infoWindowVisible: false,
      postPreviewVisible: false,
      postExpandedVisible: false,
      markers: {},
      mapOptions: null,
      activePost: null,
      activeMarkerColor: "",
      focusedPost: "",
    };
  },
  watch: {
    activePost(newPostId, oldPostId) {
      const restriction = {
        ...this.mapOptions.restriction,
        latLngBounds: viewportLatLngBounds,
      };
      // Hide old post
      if (oldPostId) {
        this.markers[oldPostId].is_post_visible = false;
      }
      // Hide current post and update map bounds to fit the viewport when click its marker
      if (!newPostId) {
        this.mapOptions = { ...this.mapOptions, restriction };
        return;
      }
      // Show post and change map bounds to default when click a marker
      this.markers[newPostId].is_post_visible = true;
      restriction.latLngBounds = defaultLatLngBounds;
      this.mapOptions = { ...this.mapOptions, restriction };
    },
  },
  created() {
    this.mapOptions = options;
  },
  async mounted() {
    await this.fetchMakers();
    this.fitBounds();
  },
  methods: {
    async fetchMakers() {
      const data = await getMarkers();
      const markers = {};

      for (let item of data) {
        markers[item.id] = { ...item.attributes, is_post_visible: false };
      }
      this.markers = markers;
    },
    handleOnClick(postId) {
      if (this.activePost === postId) {
        this.activePost = null;
        return;
      }
      this.activePost = postId;
    },
    handleOnMouseOver(postId) {
      if (postId !== this.focusedPost) {
        const PLACEHOLDER_PREFIX = "ph_";
        this.focusedPost = postId;
        this.activeMarkerColor = this.markers[postId].color;
        this.markers[postId].color =
          PLACEHOLDER_PREFIX + this.markers[postId].color;
      }
    },
    handleOnMouseOut(postId) {
      this.markers[postId].color = this.activeMarkerColor;
      this.activeMarkerColor = "";
      this.focusedPost = "";
    },
    handleOnClickMap() {
      this.activePost = null;
    },
    fitBounds() {
      this.$refs.gmap.$mapPromise.then((map) => {
        var bounds = new google.maps.LatLngBounds();
        Object.values(this.markers)
          .reverse()
          .forEach((marker) => {
            bounds.extend(
              new google.maps.LatLng(
                marker.location_point.latitude,
                marker.location_point.longitude
              )
            );
          });
        map.fitBounds(bounds);
      });
    },
    handleOnExpandPost() {
      this.postExpandedVisible = true;
    },
    handleOnClosePostExpanded() {
      this.postExpandedVisible = false;
      this.activePost = null;
    },
    zoomIn() {
      const currentZoom = this.$refs.gmap.$mapObject.zoom;

      if (currentZoom < MAX_ZOOM) {
        this.mapOptions = { ...this.mapOptions, zoom: currentZoom + 1 };
      }
    },
    zoomOut() {
      const currentZoom = this.$refs.gmap.$mapObject.zoom;

      if (currentZoom > MIN_ZOOM) {
        this.mapOptions = { ...this.mapOptions, zoom: currentZoom - 1 };
      }
    },
  },
};
</script>

<style lang="scss" scoped>
.map-view {
  height: 90vh;
  position: relative;

  .gmap-container {
    width: 100vw;
    height: 100%;
  }

  .preview-shown {
    transform: translateY(-74px);
  }

  .map-control {
    position: absolute;
    z-index: 3000;
    bottom: 48px;
    left: 12px;
    display: flex;
    flex-direction: column;

    transition: transform 0.25s cubic-bezier(0.4, 0, 0.2, 1);

    svg {
      height: 14px;
    }

    .button {
      width: 30px;
      height: 30px;
      display: flex;
      align-items: center;
      justify-content: center;
      color: #666666;
      background-color: #ffffff;
      transition: all 0.15s ease-out;
      cursor: pointer;
      border: 0;
      outline: unset;

      &:hover {
        svg {
          color: #000000;
        }
      }
    }

    .map-center-control {
      border-radius: 6px;
    }

    .group-button {
      margin-top: 6px;
      border-radius: 6px;
      overflow: hidden;
      box-shadow: 0 1px 1px rgba(0, 0, 0, 0.12);
      transition: all 0.1s cubic-bezier(0.4, 0, 0.2, 1);
    }
  }

  .slide-leave-active,
  .slide-enter-active {
    transition: 0.3s;
  }
  .slide-enter {
    transform: translate(0, 100%);
  }
  .slide-leave-to {
    transform: translate(0, 100%);
  }

  @media screen and (min-width: 768px) {
    .preview-shown {
      transform: unset;
    }

    .map-control {
      flex-direction: row;
      bottom: 26px;
      right: 12px;
      left: unset;

      svg {
        height: 14px;
      }

      .group-button {
        display: flex;
        margin-top: 0;
        margin-left: 12px;
      }
    }

    .slide-enter {
      transform: translate(100%, 0);
    }
    .slide-leave-to {
      transform: translate(100%, 0);
    }
  }
}
</style>
