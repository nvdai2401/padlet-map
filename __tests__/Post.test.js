/* eslint-disable no-undef */
import { posts } from './shares/mockData';
import { mountWithProps } from './shares/utils';
import { Post } from '@/components/MapView/components';
import {
  POST_WIDTH,
  calculatePlaceholderHeight,
  generateGmapSearchUrl,
} from '@/shares/utils';

const propsData = {
  post: posts[0],
  onExpandPost: jest.fn(),
};

describe('<Post />', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = mountWithProps(Post, propsData);
  });

  afterEach(() => {
    wrapper.destroy();
  });

  it('is a Vue instance', () => {
    expect(wrapper.isVueInstance).toBeTruthy();
  });

  it('should match snapshot', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it(`should contain post header`, () => {
    expect(wrapper.find('.post-header').text()).toBe(propsData.post.headline);
  });

  it(`should contain post content`, () => {
    expect(wrapper.find('.post-content').exists()).toBeTruthy();
  });

  it(`should contain post image`, () => {
    expect(wrapper.find('img').exists()).toBeTruthy();
  });

  it(`should call onExpandPost when click Expand post`, () => {
    wrapper.find('.more-actions-button').trigger('click');
    wrapper.vm.$nextTick(() => {
      const MoreActionsList = wrapper.find('.more-actions-list');

      MoreActionsList.findAll('li')
        .at(0)
        .trigger('click');

      expect(propsData.onExpandPost).toHaveBeenCalled();
    });
  });

  it(`should open new window when click Open in Google Map`, () => {
    wrapper.find('.more-actions-button').trigger('click');
    wrapper.vm.$nextTick(() => {
      const MoreActionsList = wrapper.find('.more-actions-list');
      const url = generateGmapSearchUrl(propsData.post.location_name);
      window.open = jest.fn();
      window.open(url, '_blank');
      MoreActionsList.findAll('li')
        .at(1)
        .trigger('click');

      expect(window.open).toHaveBeenCalledWith(url, '_blank');
    });
  });

  it(`should render image with right style and height`, () => {
    const placeholder = propsData.post.preview_image;
    const expectedHeight = calculatePlaceholderHeight(
      placeholder.width,
      placeholder.height,
      POST_WIDTH,
    );

    expect(wrapper.find('img').attributes().height).toBe(
      String(expectedHeight),
    );
    expect(wrapper.find('img').attributes().style).toBe(
      `background-color: rgb(${placeholder.dominant_color.join(', ')});`,
    );
  });
});
