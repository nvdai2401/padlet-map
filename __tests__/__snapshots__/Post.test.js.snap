// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Post /> should match snapshot 1`] = `
<div
  class="flex-column post"
>
  <img
    alt="United States"
    class="post-image"
    height="142"
    loading="lazy"
    src="https://ik.imagekit.io/fireman35/pics/usa_98nvVlwOEN.jpg?tr=w-518"
    style="background-color: rgb(102, 86, 108);"
    width="254"
  />
   
  <h3
    class="post-header"
  >
    
    United States
  
  </h3>
   
  <div
    class="post-content"
  >
    <div>
      Came here to study in 2006.
    </div>
  </div>
   
  <div
    class="flex-row align-center justify-center more-actions-button"
  >
    <svg
      aria-hidden="true"
      class="svg-inline--fa fa-ellipsis-v fa-w-6"
      data-icon="ellipsis-v"
      data-prefix="fas"
      focusable="false"
      role="img"
      viewBox="0 0 192 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        class=""
        d="M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z"
        fill="currentColor"
      />
    </svg>
  </div>
   
  <!---->
</div>
`;
